tensorflow/tools/xla_extract/tf_graph_to_xla_lib.cc

```diff
@@ -114,48 +117,65 @@ xla::HloModuleProto ExtractHloFromGraphDef(const GraphDef& in_graph,
+                                 f_.signature().name().length() - 2) == "_0");
+                   });
 
   FunctionDef fdef;
 
-  if(fdef_iter == fdef_lib.function().rend()){
-    fdef_iter = std::find_if(fdef_lib.function().rbegin(), fdef_lib.function().rend(),
-				[] (const FunctionDef& f_) -> bool {
-				  return (f_.signature().name().find("cluster_") == 0);
-				});
+  if (fdef_iter == fdef_lib.function().rend()) {
+    fdef_iter =
+        std::find_if(fdef_lib.function().rbegin(), fdef_lib.function().rend(),
+                     [](const FunctionDef& f_) -> bool {
+                       return (f_.signature().name().find("cluster_") == 0);
+                     });
   }
 
-  if(fdef_iter != fdef_lib.function().rend()){
+  if (fdef_iter != fdef_lib.function().rend()) {
     fdef = *fdef_iter;
-  }
-  else{
+  } else {
     fdef = *fdef_lib.function().begin();
-    LOG(INFO) << "cluster not found, using "<<fdef.signature().name()<<" instead\n";
+    LOG(INFO) << "cluster not found, using " << fdef.signature().name()
+              << " instead\n";
   }
 
-  // Need to rearrange these xla_args to match graph.
-  // features and labels should be the first nodes but sometimes aren't
-  // If args are [a, b, c, d, e, ...] and a is the features arg, then leave
-  // as is.  if d is the features arg, then reorder as [d, e, ..., a, b, c]
-  // This corresponds to a rotation with `d` as the axis.
   auto xla_args = BuildXlaArgsFromClientGraph(client_graph);
 
-  // Find the features arg (first parameter tensor)
-  auto features_iter = std::find_if(
-      xla_args.begin(), xla_args.end(),
-      [](const XlaCompiler::Argument& xarg) -> bool {
-        return (xarg.kind == XlaCompiler::Argument::kParameter &&
-                xarg.resource_kind == 0 && xarg.initialized == false);
-      });
+  // to make sure xla_args matches fdef
+
+  LOG(INFO) << "number of function defs:" << fdef_lib.function().size() << "\n";
+  LOG(INFO) << fdef.signature().name() << "\n";
+  LOG(INFO) << "xla args number:" << xla_args.size() << "\n";
+  LOG(INFO) << "fdef_args number:" << fdef.signature().input_arg().size()
+            << "\n";
+
+  auto fdef_ground_truth = fdef.signature().input_arg();
+  std::vector<XlaCompiler::Argument> new_xla_args(fdef_ground_truth.size());
+
+  const std::string kReadVarOpString = "readvariableop";
+  const std::string kIdentityString = "identity";
+  for (int i = 0; i < xla_args.size(); i++) {
+    std::string xla_arg_name = xla_args[i].name;
+    xla_arg_name = str_util::ArgDefCase(xla_arg_name);
+    xla_arg_name = xla_arg_name + "_0_arg";
+    xla_arg_name = str_util::Lowercase(xla_arg_name);
+    xla_arg_name = str_util::StringReplace(xla_arg_name, kReadVarOpString,
+                                           kIdentityString, true);
+    for (int j = 0; j < fdef_ground_truth.size(); j++) {
+      if (xla_arg_name == fdef_ground_truth[j].name()) {
+        new_xla_args[j] = xla_args[i];
+      }
```

Comment:
I still don't like this construct.  It seems like it could have a bad failure if you missed an arg transformation. and then you would pass a `new_xla_args` on that had a bunch of default constructed `XlaCompiler::Argument`s

maybe do a find in `fdef_ground_truth`, ensuring that you error out if you didn't find a match, store up the indices, and then just re-order `xla_args` according to that permutation.

I would also check that the ordering generated is actually a permutation, otherwise, there is the possibility that on different loop iterations you generate the same `xla_arg_name`

Reply:
the `xla_args` names wont always match with `fdef_ground_truth` (seeing this so far only in online norm) , so im also using it to prune away the args, which are not present in the `fdef_ground_truth`, ive added a check to make sure they all are matched.